drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py

```diff
@@ -0,0 +1,590 @@
+    location.
+
+    For more information, read the Passive Sonar wiki page of the mil_common repository.
+    '''
+    def __init__(self, input_mode='serial', custom_input_source=None):
+
+        self.input_mode = input_mode
+        self.load_params()
+        self.bagging = False
+
+        # TODO: update to use ros_alarms
+
+        if self.input_mode == 'bag':
+            self.input_source = _BaggedReceiverArray(self.input_src_params['bag'])
+
+        elif self.input_mode == 'serial':
+            self.input_source = _SerialReceiverArray(self.input_src_params['serial'], self.receiver_count)
+
+        elif self.input_mode == 'custom_cb':
+            if not issubclass(custom_input_source, ReceiverArrayInterface):
+                raise RuntimeError('The custom input source provided ({}) is not derived from {}'
+                    .format(type(custom_input_source), 'ReceiverArrayInterface'))
+            self.input_source = custom_input_source
+
+        else:
+            raise RuntimeError(self.input_mode + 'is not a supported input mode')
+
+        self.reset_position_estimate(None)
+
+        self.multilaterator = Multilaterator(self.receiver_locations, self.c, self.method)
+
+        self.plot_pub = rospy.Publisher('/passive_sonar/plot', Image, queue_size=1)
+        self.rviz_pub = rospy.Publisher("/passive_sonar/rviz", Marker, queue_size=10)
+        self.declare_services()
+        rospy.loginfo('Passive sonar driver initialized')
+
+    # Passive Sonar Helpers
+
+    def load_params(self):
+        '''
+        Loads all the parameters needed for receiving and processing signals from the passive
+        sonar board and calculating headings towards an active pinger.
+
+        These parameters are descrived in detail in the Passive Sonar page of the mil_common wiki.
+        TODO: copy url here
+        '''
+        # ROS params expected to be loaded under namespace passive_sonar
+        self.required_params = ['receiver_locations', 'method', 'c', 'target_frequency',
+                                'sampling_frequency', 'upsampling_factor', 'locating_frame',
+                                'receiver_array_frame', 'min_variance', 'observation_buffer_size',
+                                'input_timeout']
+        self.input_src_params = \
+        {
+         'serial' : ['port', 'baud', 'tx_request_code', 'tx_start_code', 'read_timeout',
+                     'scalar_size', 'signal_size', 'signal_bias', 'locating_frame',
+                     'receiver_array_frame'],
+         'bag'    : ['bag_filename' ,'locating_frame', 'receiver_array_frame']
+        }
+
+        load = lambda prop: setattr(self, prop, rospy.get_param('passive_sonar/' + prop))
+        try:
+            [load(x) for x in self.required_params]
+        except KeyError as e:
+            raise IOError('A required rosparam was not declared: ' + str(e))
+
+        self.receiver_count = len(self.receiver_locations) + 1
```

Comment:
Wouldn't having to specify all `receiver_locations` make more sense, does one of the hydrophones _have_ to be the origin? (Assuming the +1 is because you don't have to specify the (0, 0, 0) center hydrophone which may be bad assumption)